_posts/2023-04-21-tensorflow-mnist-classification.adoc

```diff
@@ -0,0 +1,405 @@
+----
+
+This step may take quite some time to finish.
+However, once finished, our model is ready to recognize handwritten digits!
+
+=== Streaming the data into the model
+
+Let's see how good our model is in digit recognition.
+But as our main goal here is to explore the means how to ingest data into TensorFlow, we will start model evaluation on empty (or more accurately even non-existing) Kafka topic and see if we will be able to evaluate the data on the fly as they will pop-up first in the database and then in the corresponding Kafka topic.
+For this purpose we can use one of the above-mentioned streaming classes:
+
+[source,python]
+----
+test_ds = tfio.experimental.streaming.KafkaGroupIODataset(
+    topics=[KAFKA_TEST_TOPIC],
+    group_id=KAFKA_CONSUMER_GROUP,
+    servers=KAFKA_SERVERS,
+    stream_timeout=9000,
+    configuration=[
+        "session.timeout.ms=10000",
+        "max.poll.interval.ms=10000",
+        "auto.offset.reset=earliest"
+    ],
+)
+----
+
+Again, arguments are mostly self-explanatory.
+Two things may need further explanation: `stream_timeout` and `configuration` parameters.
+`stream_timeout` determines interval of inactivity (in milliseconds) after which the streaming would terminate.
+`configuration` is https://github.com/confluentinc/librdkafka/blob/master/CONFIGURATION.md[librdkafka configuration].
+It's basically a configuration of Kafka client and you should configure at least session timeout (`session.timeout.ms`) and its poll interval (`max.poll.interval.ms`).
+The values of these parameters should be higher than the value of `stream_timeout`.
+
+The dataset provided by this loader is slightly different - instead of providing a single record containing message and its key, we get key and message already split.
+Therefore, we have to define slightly modified map function with two arguments:
+
+[source,python]
+----
+def decode_kafka_stream_record(message, key):
+    img_int = tf.io.decode_csv(message, [[0.0] for i in range(NUM_COLUMNS)])
+    img_norm = tf.cast(img_int, tf.float32) / 255.
+    label_int = tf.strings.to_number(key, out_type=tf.dtypes.int32)
+    return (img_norm, label_int)
+----
+
+With this function, we can adjust the dataset and create batches as before:
+
+[source,python]
+----
+test_ds = test_ds.map(decode_kafka_stream_record)
+test_ds = test_ds.batch(BATCH_SIZE)
+----
+
+and evaluate the model:
+
+[source,python]
+----
+model.evaluate(test_ds)
+----
+
+You can execute a cell with model evaluation in the Jupyter notebook.
+As there is no such topic in Kafka and even no table with test data in the database, it will wait.
+However, as the streaming timeout is set to 9 seconds, after launching the model evaluation you have to provide some data within this time frame.
+You can use a prepared SQL file, which was created at the very beginning of this demo.
+In `postgres` directory, there is `mnist_test.sql` file, which can be easily used for creating test data sample the database:
```

Comment:
```suggestion
The execution will wait because there is no such topic in Kafka and no table with test data in the database.
The streaming timeout is 9 seconds, so data must be provided within this time frame after launching the model evaluation.
At the start of this demo, we created a SQL file in the `postgres` directory called `mnist_test.sql`, which can generate the test data we need:
```